File: cpp/0135分发糖果.cpp
#include<vector>
using namespace std;

int candy(vector<int> &ratings)
{
    const int len = ratings.size();
    int allocations[len];

    for (int i = 0; i < len; i++)
    {
        allocations[i] = 1;
    }

    for (int i = 1; i < len; i++)
    {
        if(ratings[i]>ratings[i-1])
            allocations[i] = allocations[i - 1] + 1;
    }

    for (int i = len - 2; i >= 0;i--)
    {
        if(ratings[i]>ratings[i+1])
            allocations[i] = max(allocations[i], allocations[i + 1] + 1);
    }

    int sum = 0;
    for (int i : allocations)
    {
        sum += i;
    }

    return sum;
}

File: python/0135分发糖果.py

def candy(ratings: list) -> int:
    size=len(ratings)
    allocation=[1]*size

    # 第一次贪心从左向右,如果比左面的rating大就比左面多分一个
    for i in range(1,size):
        if(ratings[i]>ratings[i-1]):
            allocation[i]=allocation[i-1]+1

    # 第二贪心从右向左，同时兼顾左右
    for i in range(size-1,0,-1):
        if(ratings[i-1]>ratings[i]):
            allocation[i-1]=max(allocation[i]+1,allocation[i-1])
    
    return sum(allocation)

print(candy([1,3,2,2,1]))